### PR TITLE
feat(sub): Indicate resolution conflicts on Subscription statuses

### PR DESCRIPTION
**Pull request by anik120**
**Description of the change:**

Resolution considers everything in a given namespace as a unit, so conflicts aren't
generated on a per-Subscription basis. However, as the main entry point to resolution,
users first look at the Subscription to understand resolution failures. The set of conflicts
today is written to Events with type ResolutionFailed. This PR projects resolution errors onto
all subscriptions on the namespace for any resolution failure.

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

**Comment by anik120** [comment]
/hold

Made changes to the subscription api inside the vendor folder for reviews. PR for api change is here: git push --force-with-lease origin sub-resolution-error

will unhold once this and the api repo PR is lgtmed and changes from api is vendored in here.

File: pkg/controller/operators/catalog/operator.go [hunk 1]
```diff
@@ -909,9 +909,22 @@ func (o *Operator) syncResolvingNamespace(obj interface{}) error {
...
+				logger.WithError(updateErr).Debug("failed to update subs conditions")
+			}
 			return nil
 		}
+		updateErr := o.setSubsResFailedCond(subs, "Dependency resolution failed", err.Error(), true)
+		if updateErr != nil {
+			logger.WithError(updateErr).Debug("failed to update subs conditions")
+		}
 		return err
+	} else {
```
Comment (anik120) [inline pkg/controller/operators/catalog/operator.go:923]: Question: This is under the assumption that if there is a terminal resolution error, user intervention can resolve the error and get the subscriptions in the objects back to a good state. Is that a correct assumption?

File: test/e2e/subscription_e2e_test.go [hunk 2]
```diff
@@ -2105,6 +2105,55 @@ var _ = Describe("Subscription", func() {
...
+		})
+
+		AfterEach(func() {
+			cleanup()
+			teardown()
+		})
+
+		It("the subscription has a condition in it's status that indicates the resolution error", func() {
+			_, err := fetchSubscription(crc, testNamespace, subName, subscriptionHasResolutionFailedCondChecker)
+			require.NoError(GinkgoT(), err)
```
Comment (njhale) [inline test/e2e/subscription_e2e_test.go:2152]: If we can, we should directly use ginko/gomega for net-new tests -- something like:

```go
Eventually(func() (corev1.ConditionStatus, error) {
  sub, err = crc.OperatorsV1alpha1().Subscriptions(namespace).Get(context.Background(), name, metav1.GetOptions{})
  if err != nil {
    return corev1.ConditionUnknown, err
  }
  return sub.Status.GetCondition(operatorsv1alpha1.SubscriptionResolutionFailed), nil
}).Should(Equal(corev1.ConditionTrue))
```

this keeps us from building deeper into our old homespun test frameworks.

File: test/e2e/subscription_e2e_test.go [hunk 3]
```diff
@@ -2488,6 +2537,20 @@ func subscriptionStateUpgradeAvailableChecker(subscription *operatorsv1alpha1.Su
...
 }
 
+func subscriptionHasResolutionFailedCondChecker(subscription *operatorsv1alpha1.Subscription) bool {
+	if cond := subscription.Status.GetCondition(operatorsv1alpha1.SubscriptionResolutionFailed); cond.Status == corev1.ConditionTrue {
+		return true
+	}
+	return false
+}
+
+func subscriptionDoesNotHaveResolutionFailedCondChecker(subscription *operatorsv1alpha1.Subscription) bool {
```
Comment (njhale) [inline test/e2e/subscription_e2e_test.go:2547]: looks like this function isn't being used anywhere

**Review by njhale — COMMENTED** [review]
Looking great so far.

File: pkg/controller/operators/catalog/operator.go [hunk 4]
```diff
@@ -1189,6 +1202,55 @@ func (o *Operator) createInstallPlan(namespace string, gen int, subs []*v1alpha1
 	return reference.GetReference(res)
 }
 
+func (o *Operator) setSubsResFailedCond(subs []*v1alpha1.Subscription, reason, message string, setTrue bool) error {
```
[the remark this answers is not in the document]
Comment (anik120) [reply pkg/controller/operators/catalog/operator.go:1205]: It's being set to `false` here https://github.com/operator-framework/operator-lifecycle-manager/pull/2269/commits/bc1d24328c4ceeca4901d00c94e3890bc0f278b4#diff-a1760d9b7ac1e93734eea675d8d8938c96a50e995434b163c6f77c91bace9990R924

File: test/e2e/subscription_e2e_test.go (hunk 3 above)
Comment (anik120) [reply test/e2e/subscription_e2e_test.go:2547]: Removing both of these functions while changing the test according to https://github.com/operator-framework/operator-lifecycle-manager/pull/2269#discussion_r674067049

File: pkg/controller/operators/catalog/operator.go (hunk 1 above)
Comment (njhale) [reply pkg/controller/operators/catalog/operator.go:923]: cluster admins can always nuke everything and recreate after the fact. there are also (rather gnarly) workarounds to pop an InstallPlan back into a non-terminal state for a retry (e.g. manually patch InstallPlan status).

**Review by njhale — COMMENTED** [review]
Nice!

My only real concern is in the lack of unit tests.

**Comment by anik120** [comment]
@njhale  there are tests [here](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator_test.go#L1071) that tests that the method returns the correct value/error when resolver error occurs.
Example the [NotSatisfiableError test](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator_test.go#L1117) checks that the sovler.NotSatisfiable error is returned when a constraint is not satisfyiable. And there's [another unit test](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator_test.go#L1181) to test any resolver error that is not a constraint not satisfyiable error. So we have unit tests that covers all scenarios for resolver error during namespace sync, this PR just bubbles the errors up to the subscriptions, which needed the e2e test to test

**Comment by njhale** [comment]
> @njhale there are tests [here](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator_test.go#L1071) that tests that the method returns the correct value/error when resolver error occurs.
> Example the [NotSatisfiableError test](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator_test.go#L1117) checks that the sovler.NotSatisfiable error is returned when a constraint is not satisfyiable. And there's [another unit test](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/catalog/operator_test.go#L1181) to test any resolver error that is not a constraint not satisfyiable error. So we have unit tests that covers all scenarios for resolver error during namespace sync, this PR just bubbles the errors up to the subscriptions, which needed the e2e test to test

hmm, I was thinking of the class of test that uses a mock client and tests a wider range of I/O (i.e. not just checking the error) -- but since the e2e test is so similar, I don't think it's worth holding so close to freeze.

/approve

File: pkg/controller/operators/catalog/operator.go [hunk 5]
```diff
@@ -909,9 +909,22 @@ func (o *Operator) syncResolvingNamespace(obj interface{}) error {
 		// not-satisfiable error
 		if _, ok := err.(solver.NotSatisfiable); ok {
 			logger.WithError(err).Debug("resolution failed")
+			updateErr := o.setSubsCond(subs, v1alpha1.SubscriptionResolutionFailed, "Constraint not satisfyable", err.Error(), true)
```
Comment (kevinrizza) [inline pkg/controller/operators/catalog/operator.go:912]: nit: spelling error. this should say `Constraint not satisfiable`

**Comment by kevinrizza** [comment]
minus that one nit, this looks good to me

File: pkg/controller/operators/catalog/operator.go [hunk 6]
```diff
@@ -909,9 +909,22 @@ func (o *Operator) syncResolvingNamespace(obj interface{}) error {
 		// not-satisfiable error
 		if _, ok := err.(solver.NotSatisfiable); ok {
 			logger.WithError(err).Debug("resolution failed")
+			updateErr := o.setSubsCond(subs, v1alpha1.SubscriptionResolutionFailed, "Constraint not satisfiable", err.Error(), true)
```
Comment (benluddy) [inline pkg/controller/operators/catalog/operator.go:912]: Isn't Reason usually intended to be machine-readable?

Comment (kevinrizza) [reply pkg/controller/operators/catalog/operator.go]: nice catch

**Pull request merged**